File: fast_and_slow_pointer/middle-of-a-linkedlist.js
class Node {
    constructor(value, next = null) {
        this.value = value;
        this.next = next;
    }
}

const find_middle_of_linked_list = function (head) {
    let l = head;
    let r = head.next;

    while (r) {
        r = r.next?.next;
        l = l.next;
    }

    return l;
}

head = new Node(1);
head.next = new Node(2);
head.next.next = new Node(3);
head.next.next.next = new Node(4);
head.next.next.next.next = new Node(5);

console.log(`LinkedList cycle start: ${find_middle_of_linked_list(head).value}`); // 3

head.next.next.next.next.next = new Node(6);
console.log(`LinkedList cycle start: ${find_middle_of_linked_list(head).value}`); // 4

head.next.next.next.next.next.next = new Node(7);
console.log(`LinkedList cycle start: ${find_middle_of_linked_list(head).value}`); // 4

File: fast_and_slow_pointer/python/middle-of-a-linkedlist.py
class Node:
    def __init__(self, value, next=None):
        self.value = value
        self.next = next


def find_middle_of_linked_list(head):
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
    return slow


if __name__ == '__main__':
    head = Node(1)
    head.next = Node(2)
    head.next.next = Node(3)
    head.next.next.next = Node(4)
    head.next.next.next.next = Node(5)
    print(f"Middle Node: {find_middle_of_linked_list(head).value}")

    head.next.next.next.next.next = Node(6)
    print(f"Middle Node: {find_middle_of_linked_list(head).value}")

    head.next.next.next.next.next.next = Node(7)
    print(f"Middle Node: {find_middle_of_linked_list(head).value}")
